### ISSUE TITLE
FR: customizable grain timeout

### ISSUE THREAD

**Issue by abliss** [sandstorm-io/sandstorm#3297]
[Currently](https://github.com/sandstorm-io/sandstorm/blob/master/src/sandstorm/supervisor.c%2B%2B#L547-L548), the timeout for a grain in the absence of keepalives is hardcoded at 90 seconds. When developing and debugging, I often seem to get my grain killed out from under me, despite leaving a tab open and pointing to it. It would be nice for a way (at least inside `spk dev`) to make a grain never timeout (or timeout after some much longer delay).

**Comment by ocdtrekkie** [comment]
Is the tab active in your browser? Browsers suspend tabs for performance reasons, which might lead to a premature kill. A different browser window though should be a safe choice.

**Comment by zenhack** [comment]
Yes, this is part of the problem, and the other night we worked around
it by finding somewhere to pin a tab so that it wouldn't get
deactivated, but I think it would still be an improvement to the
developer experience to have an easy way to just say "don't kill this
grain."

Quoting Jacob Weisz (2020-04-25 18:05:40)

>    Is the tab active in your browser? Browsers suspend tabs for
>    performance reasons, which might lead to a premature kill. A different
>    browser window though should be a safe choice.

**Comment by abliss** [comment]
I definitely saw at least one grain kill while the tab was open and live on
my screen.

**Comment by ocdtrekkie** [comment]
That in particular seems like it'd be a significant bug, unless we can discern some browser behavior that caused it.

**Comment by zenhack** [comment]
I'd be in favor of adding a mechanism to mark a dev-mode grain as "don't kill." Perhaps we can do this as a meteor method, so you could pop open the browser console and do something like:

```javascript
Meteor.call('setGrainKeepalive, true)
```

To tell sandstorm not to kill the currently active grain.

I think this would be relatively easy to implement.

**Comment by zenhack** [comment]
Also: the 90-second hardcoded value is how often the supervisor checks if it's been told to stay running. Probably the way the above would work is just by calling Supervisor.keepAlive() more often.

**Comment by zenhack** [comment]
Note that if the page isn't actively making network calls, then I don't think having the browser open is sufficient to keep the server alive. The relevant bit is if the supervisor sees activity, which doesn't need to happen if the page is rendered and the grain's server isn't being engaged.

**Comment by ocdtrekkie** [comment]
Honestly, my assumption was that having the browser open was indeed holding the grain open! Is the way that you think it currently works ideal, or is that in itself problematic?

**Comment by zenhack** [comment]
It hasn't bit me in any meaningful way. I don't expect the consequences would be any worse than the same slowness we see for some apps on grain startup. It's not obvious to me that it's a problem outside of a dev context.

**Comment by kentonv** [comment]
> Note that if the page isn't actively making network calls, then I don't think having the browser open is sufficient to keep the server alive.

That's not supposed to be true... as long as the client is subscribing to the session to keep it alive, we're supposed to be calling keepAlive() on the grain's supervisor periodically.

However, now that I look, I cannot find the code that does this. I am very confused. Maybe it got lost when I rewrote the gateway to C++? But I can't believe no one has noticed the problem!

Either `bumpSession()` in `gateway-router.js` or `updateLastActive()` in `backend.js` (which `bumpSession()` calls) really ought to be invoking `keepAlive` on the supervisor...

OTOH, if we've gotten this far and this hasn't been a major issue, I wonder if this suggests that letting grains die if they aren't getting actual traffic is a good idea after all.
